src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs

```diff
@@ -3,17 +3,129 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics.Contracts;
+using System.Runtime.CompilerServices;
 
 namespace System.Dynamic.Utils
 {
     internal static partial class CollectionExtensions
     {
+        public static T[] AddFirst<T>(this IList<T> list, T item)
+        {
+            T[] res = new T[list.Count + 1];
+            res[0] = item;
+            list.CopyTo(res, 1);
+            return res;
+        }
+
         public static T[] AddLast<T>(this IList<T> list, T item)
         {
             T[] res = new T[list.Count + 1];
             list.CopyTo(res, 0);
             res[list.Count] = item;
             return res;
         }
+
+        public static T[] RemoveFirst<T>(this T[] array)
+        {
+            T[] result = new T[array.Length - 1];
+            Array.Copy(array, 1, result, 0, result.Length);
+            return result;
+        }
+
+        public static T[] RemoveLast<T>(this T[] array)
+        {
+            T[] result = new T[array.Length - 1];
+            Array.Copy(array, 0, result, 0, result.Length);
+            return result;
+        }
+
+        // Name needs to be different so it doesn't conflict with Enumerable.Select
+        public static U[] Map<T, U>(this T[] array, Func<T, U> select)
```

Comment:
Does the delegate type matter for usage of Map?  I ask because if it doesn't, since this method is internal, it could be removed entirely and replaced with usage of Array.ConvertAll, which does the same thing with the same signature, other than the type of the delegate.